File: src/rt/env/array.js
(function () {
	'use strict';
	
	var isCallable = Intrinsics.isCallable,
	    hasProperty = Intrinsics.hasProperty,
	    registerFunction = Intrinsics.registerFunction;

	// 15.4.4.2 Array.prototype.toString ( )
	registerFunction(Array.prototype, function toString() {
		var array = (object)this;
		
		var fn = array.join;
		if (!isCallable(fn)) {
			fn = Object.prototype.toString;
		}
		
		return fn.call(array);
	});
	
	// 15.4.4.3 Array.prototype.toLocaleString ( )
	registerFunction(Array.prototype, function toLocaleString() {
		var array = (object)this;
		var arrayLen = array.length;
		var len = (u32)arrayLen;
		
		if (len == 0) {
			return '';
		}
		
		var result = '';
		
		for (var i = 0; i < len; i++) {
			if (i > 0) {
				result += ',';
			}
			
			var element = array[i];
			
			if (typeof element !== 'null' && typeof element !== 'undefined') {
				element = (object)element;
				
				var toLocaleString = element.toLocaleString;
				if (!isCallable(toLocaleString)) {
					throw new TypeError('Cannot call toLocaleString on array element');
				}
				
				result += toLocaleString.call(element);
			}
		}
	});
	
	// 15.4.4.4 Array.prototype.concat ( [ item1 [ , item2 [ , … ] ] ] )
	registerFunction(Array.prototype, function concat(item1) {
		function append(element) {
			if (Array.isArray(element)) {
				for (var i = 0; i < element.length; i++) {
					if (hasProperty(element, i)) {
						defineOwnProperty(result, offset, element[i]);
					}
					
					offset++;
				}
			} else {
				defineOwnProperty(result, offset++, element);
			}
		}
		
		var array = (object)this;
		var result = [];
		var offset = 0;
		
		append(array);
		
		for (var i = 0; i < arguments.length; i++) {
			append(arguments[i]);
		}
		
		// TODO: This is not conform the spec and covers the following scenario:
		//
		//   var x = [];
		//   x.length = 10;
		//   var y = [].concat(x);
		//   assert(y.length == 10);
		//
		// The problem is that when elements are missing (and [[HasOwnProperty]] is
		// supposed to return false), the spec concat implementation does not
		// update the length of the result array.
		result.length = offset;
		
		return result;
	});
	
	// 15.4.4.5 Array.prototype.join (separator)
	registerFunction(Array.prototype, function join(separator) {
		var array = (object)this;
		var len = (u32)array.length;
		
		if (typeof separator === 'undefined') {
			separator = ',';
		} else {
			separator = (string)separator;
		}
		
		if (len == 0) {
			return '';
		}
		
		var result = '';
		
		for (var i = 0; i < len; i++) {
			if (i > 0) {
				result += separator;
			}
			
			var element = array[i];
			if (typeof element !== 'null' && typeof element !== 'undefined') {
				result += (string)element;
			}
		}
		
		return result;
	});
	
	
	// 15.4.4.8 Array.prototype.reverse ( )
	registerFunction(Array.prototype, function reverse() {
		var array = (object)this;
		var len = (u32)array.length;
		
		var middle = Math.floor(len / 2);
		var lower = 0;
		
		while (lower != middle) {
			var upper = len - lower - 1;
			
			var lowerValue = array[lower];
			var upperValue = array[upper];
			
			var lowerExists = hasProperty(array, lower);
			var upperExists = hasProperty(array, upper);
			
			if (lowerExists && upperexists) {
				array[lower] = upperValue;
				array[upper] = lowerValue;
			} else if (!lowerExists && upperExists) {
				array[lower] = upperValue;
				delete array[upper];
			} else if (lowerExists && !upperExists) {
				delete[lower];
				array[upper] = lowerValue;
			}
		}
	});
	
	// 15.4.4.10 Array.prototype.slice (start, end)
	registerFunction(Array.prototype, function slice(start, end) {
		var array = (object)this;
		var result = [];
		
		var len = (u32)array.length;
		var start = (int)start;
		
		var offset;
		if (start < 0) {
			offset = Math.max(len + start, 0);
		} else {
			offset = Math.min(start, len);
		}
		
		if (typeof end === 'undefined') {
			end = len;
		} else {
			end = (int)end;
		}
		
		var last;
		if (end < 0) {
			last = Math.max(len + end, 0);
		} else {
			last = Math.min(end, len);
		}
		
		for (; offset < last; offset++) {
			if (hasProperty(array, offset)) {
				result.push(array[offset]);
			}
		}
	});
	
	
	// 15.4.4.12 Array.prototype.splice (start, deleteCount [ , item1 [ , item2 [ , … ] ] ] )
	registerFunction(Array.prototype, function splice(start, deleteCount) {
		var i, k;

		var array = (object)this;
		var result = [];
		
		var len = (u32)array.length;
		
		var start = (int)start;
		if (start < 0) {
			start = Math.max(len + start, 0);
		} else {
			start = Math.min(start, len);
		}
		
		deleteCount = Math.min(Math.max((int)deleteCount, 0), len - start);
		
		for (i = 0; i < deleteCount; i++) {
			var from = start + i;
			
			if (hasProperty(array, from)) {
				result[i] = array[from];
			}
		}
		
		var itemCount = arguments.length - 2;
		
		if (itemCount < deleteCount) {
			for (k = start; k < len - deleteCount; k++) {
				var from = k + deleteCount;
				var to = k + itemCount;
				
				if (Instrinsics.hasProperty(array, from)) {
					array[to] = array[from];
				} else {
					delete array[to];
				}
			}
			
			for (k = len; k > len - deleteCount + itemCount; k--) {
				delete array[k - 1];
			}
		} else if (itemCount > deleteCount) {
			for (k = len - deleteCount; k > start; k--) {
				var from = k + deleteCount - 1;
				var to = k + itemCount - 1;
				
				if (hasProperty(array, from)) {
					array[to] = array[from];
				} else {
					delete array[to];
				}
			}
		}
		
		k = start;
		for (i = 2; i < arguments.length; i++) {
			array[k++] = arguments[i];
		}
		
		array.length = len - deleteCount + itemCount;
		
		return result;
	});
	
	// 15.4.4.16 Array.prototype.every ( callbackfn [ , thisArg ] )
	registerFunction(Array.prototype, function every(callback) {
		var thisArg = arguments[1];
		
		var array = (object)this;
		var len = (u32)array.length;
		
		if (!isCallable(callback)) {
			throw new TypeError('Callback must be callable');
		}
		
		for (var i = 0; i < len; i++) {
			if (hasProperty(array, i)) {
				var result = callback.call(thisArg, array[i], i, array);
				if (!(bool)result) {
					return false;
				}
			}
		}
		
		return true;
	});
	
	// 15.4.4.17 Array.prototype.some ( callbackfn [ , thisArg ] )
	registerFunction(Array.prototype, function some(callback) {
		var thisArg = arguments[1];
		
		var array = (object)this;
		var len = (u32)array.length;
		
		if (!isCallable(callback)) {
			throw new TypeError('Callback must be callable');
		}
		
		for (var i = 0; i < len; i++) {
			if (hasProperty(array, i)) {
				var result = callback.call(thisArg, array[i], i, array);
				if ((bool)result) {
					return true;
				}
			}
		}
		
		return false;
	});
	
	
	// 15.4.4.18 Array.prototype.forEach ( callbackfn [ , thisArg ] )
	registerFunction(Array.prototype, function forEach(callback) {
		var thisArg = arguments[1];
		
		var array = (object)this;
		var len = (u32)array.length;
		
		if (!isCallable(callback)) {
			throw new TypeError('Callback must be callable');
		}
		
		for (var i = 0; i < len; i++) {
			if (hasProperty(array, i)) {
				callback.call(thisArg, array[i], i, array);
			}
		}
		
		return undefined;
	});
	
	// 15.4.4.19 Array.prototype.map ( callbackfn [ , thisArg ] )
	registerFunction(Array.prototype, function map(callback) {
		var thisArg = arguments[1];
		
		var array = (object)this;
		var len = (u32)array.length;
		var result = [];
		
		if (!isCallable(callback)) {
			throw new TypeError('Callback must be callable');
		}
		
		for (var i = 0; i < len; i++) {
			if (hasProperty(array, i)) {
				result.push(callback.call(thisArg, array[i], i, array));
			}
		}
		
		return undefined;
	});
	
	function defineOwnProperty(target, index, value) {
		Object.defineProperty(target, index, {
			value: value,
			writable: true,
			enumberable: true,
			configurable: true
		});
	}
	
	// 15.4.4.20 Array.prototype.filter ( callbackfn [ , thisArg ] )
	registerFunction(Array.prototype, function filter(callback) {
		var thisArg = arguments[1];
		
		var array = (object)this;
		var len = (u32)array.length;
		var result = [];
		
		if (!isCallable(callback)) {
			throw new TypeError('Callback must be callable');
		}
		
		for (var i = 0; i < len; i++) {
			if (hasProperty(array, i)) {
				var value = array[i];
				var selected = callback.call(thisArg, value, i, array);
				if ((bool)selected) {
					defineOwnProperty(result, result.length, value);
				}
			}
		}
		
		return result;
	});
	
	// 15.4.4.21 Array.prototype.reduce ( callbackfn [ , initialValue ] )
	registerFunction(Array.prototype, function reduce(callback) {
		var array = (object)this;
		var len = (u32)array.length;
		
		if (!isCallable(callback)) {
			throw new TypeError('Callback must be callable');
		}
		
		var offset = 0;
		
		var accumulator;
		if (arguments.length > 1) {
			accumulator = arguments[1];
		} else {
			for (; offset < len; offset++) {
				if (hasProperty(array, offset)) {
					accumulator = array[offset];
					break;
				}
			}
			
			if (offset == len) {
				throw new TypeError('Initial value not specified and array is empty');
			}
		}
		
		for (; offset < len; offset++) {
			if (hasProperty(array, offset)) {
				var value = array[offset];
				accumulator = callback.call(undefined, accumulator, value, offset, array);
			}
		}
		
		return accumulator;
	});
	
	// 15.4.4.22 Array.prototype.reduceRight ( callbackfn [ , initialValue ] )
	registerFunction(Array.prototype, function reduceRight(callback) {
		var array = (object)this;
		var len = (u32)array.length;
		
		if (!isCallable(callback)) {
			throw new TypeError('Callback must be callable');
		}
		
		if (len == 0 && arguments.length < 2) {
			throw new TypeError('Initial value not specified and array is empty');
		}
		
		var offset = len - 1;
		var accumulator;
		
		if (arguments.length > 1) {
			accumulator = arguments[1];
		} else {
			for (; offset >= 0; offset--) {
				if (hasProperty(array, offset)) {
					accumulator = array[offset];
					break;
				}
			}
			
			if (offset < 0) {
				throw new TypeError('Initial value not specified and array is empty');
			}
		}
		
		for (; offset >= 0; offset--) {
			if (hasProperty(array, offset)) {
				var value = array[offset];
				accumulator = callback.call(undefined, accumulator, value, offset, array);
			}
		}
		
		return accumulator;
	});
})();

File: src/rt/env/array.rs
#![allow(unused_variables)]

use ::{JsResult, JsError};
use rt::{JsEnv, JsArgs, JsValue, JsFnMode, JsItem, JsDescriptor, JsType};
use gc::*;
use syntax::Name;
use syntax::token::name;
use std::cmp::Ordering;
use std::u32;

macro_rules! local_try {
	( $expr:expr, $error:ident ) => {
		match $expr {
			Ok(ok) => ok,
			Err(error) => {
				$error = Some(error);
				return Ordering::Equal;
			}
		}
	}
}

// 15.4.1 The Array Constructor Called as a Function
// 15.4.2.1 new Array ( [ item0 [ , item1 [ , … ] ] ] )
// 15.4.2.2 new Array (len)
pub fn Array_constructor(env: &mut JsEnv, mode: JsFnMode, args: JsArgs) -> JsResult<Local<JsValue>> {
	if !mode.construct() {
		let target_args = args.args(env);
		return args.function(env).construct(env, target_args);
	}
	
	let mut array = env.create_array();
	
	if args.argc == 1 {
		let arg = args.arg(env, 0);
		if arg.ty() == JsType::Number {
			if let Some(len) = try!(arg.to_uint32_exact(env)) {
				let len = env.new_number(len as f64);
				try!(array.put(env, name::LENGTH, len, false));
			} else {
				return Err(JsError::new_range(env));
			}
		} else {
			try!(array.define_own_property(env, Name::from_index(0), JsDescriptor::new_simple_value(arg), false));
		}
	} else {
		for i in 0..args.argc {
			let arg = args.arg(env, i);
			try!(array.define_own_property(env, Name::from_index(i), JsDescriptor::new_simple_value(arg), false));
		}
	}
	
	Ok(array.as_value(env))
}

// 15.4.4.6 Array.prototype.pop ( )
pub fn Array_pop(env: &mut JsEnv, _mode: JsFnMode, args: JsArgs) -> JsResult<Local<JsValue>> {
	let mut array = try!(args.this(env).to_object(env));
	let len_val = try!(array.get(env, name::LENGTH));
	let len = try!(len_val.to_uint32(env));
	
	if len == 0 {
		let length = env.new_number(0f64);
		try!(array.put(env, name::LENGTH, length, true));
		
		Ok(env.new_undefined())
	} else {
		let index = len as usize - 1;
		let element = try!(array.get(env, Name::from_index(index)));
		try!(array.delete(env, Name::from_index(index), true));
		
		let length = env.new_number(index as f64);
		try!(array.put(env, name::LENGTH, length, true));
		
		Ok(element)
	}
}

// 15.4.4.7 Array.prototype.push ( [ item1 [ , item2 [ , … ] ] ] )
pub fn Array_push(env: &mut JsEnv, _mode: JsFnMode, args: JsArgs) -> JsResult<Local<JsValue>> {
	let mut array = try!(args.this(env).to_object(env));
	let len_val = try!(array.get(env, name::LENGTH));
	let mut index = try!(len_val.to_uint32(env)) as usize;
	
	for i in 0..args.argc {
		let arg = args.arg(env, i);
		
		let name = if index >= u32::MAX as usize {
			env.intern(&index.to_string())
		} else {
			Name::from_index(index)
		};
		
		try!(array.put(env, name, arg, true));
		
		index += 1;
	}
	
	let length = env.new_number(index as f64);
	try!(array.put(env, name::LENGTH, length, true));
	
	Ok(length)
}
}

// 15.4.4.9 Array.prototype.shift ( )
pub fn Array_shift(env: &mut JsEnv, _mode: JsFnMode, args: JsArgs) -> JsResult<Local<JsValue>> {
	let mut array = try!(args.this(env).to_object(env));
	let len_val = try!(array.get(env, name::LENGTH));
	let len = try!(len_val.to_uint32(env)) as usize;
	
	let result = if len == 0 {
		let length = env.new_number(0f64);
		try!(array.put(env, name::LENGTH, length, true));
		
		env.new_undefined()
	} else {
		let first = try!(array.get(env, Name::from_index(0)));
		
		for i in 1..len {
			let to = i - 1;
			let from_present = array.has_property(env, Name::from_index(i));
			
			if from_present {
				let from_val = try!(array.get(env, Name::from_index(i)));
				try!(array.put(env, Name::from_index(to), from_val, true));
			} else {
				try!(array.delete(env, Name::from_index(to), true));
			}
		}
		
		try!(array.delete(env, Name::from_index(len - 1), true));
		let length = env.new_number((len - 1) as f64);
		try!(array.put(env, name::LENGTH, length, true));
		
		first
	};
	
	Ok(result)
}

// 15.4.4.10 Array.prototype.slice (start, end)
pub fn Array_slice(env: &mut JsEnv, _mode: JsFnMode, args: JsArgs) -> JsResult<Local<JsValue>> {
	let array = try!(args.this(env).to_object(env));
	let mut result = env.create_array();
	let len_val = try!(array.get(env, name::LENGTH));
	let len = try!(len_val.to_uint32(env)) as usize;
	
	let relative_start = try!(args.arg(env, 0).to_integer(env));
	let mut k = if relative_start < 0f64 {
		let k = len as f64 + relative_start;
		if k < 0f64 { 0 } else { k as usize }
	} else {
		if relative_start < len as f64 { relative_start as usize } else { len }
	};
	
	let end = args.arg(env, 1);
	let relative_end = if end.is_undefined() {
		len as f64
	} else {
		try!(end.to_integer(env))
	};
	
	let final_ = if relative_end < 0f64 {
		let final_ = len as f64 + relative_end;
		if final_ > 0f64 { final_ as usize } else { 0 }
	} else {
		if relative_end < len as f64 { relative_end as usize } else { len }
	};
	
	let mut n = 0;
	
	while k < final_ {
		let k_present = array.has_property(env, Name::from_index(k));
		if k_present {
			let k_value = try!(array.get(env, Name::from_index(k)));
			try!(result.define_own_property(env, Name::from_index(n), JsDescriptor::new_simple_value(k_value), false));
		}
		
		k += 1;
		n += 1;
	}
	
	Ok(result.as_value(env))
}

// 15.4.4.11 Array.prototype.sort (comparefn)
// TODO: This is not a correct implementation!
pub fn Array_sort(env: &mut JsEnv, _mode: JsFnMode, args: JsArgs) -> JsResult<Local<JsValue>> {
	let mut obj = try!(args.this(env).to_object(env));
	let len_val = try!(obj.get(env, name::LENGTH));
	let len = try!(len_val.to_uint32(env)) as usize;
	
	let mut values = Vec::new();
	
	for i in 0..len {
		if obj.has_property(env, Name::from_index(i)) {
			values.push(Some(try!(obj.get(env, Name::from_index(i)))));
		} else {
			values.push(None)
		}
	}
	
	let compare_fn = {
		let arg = args.arg(env, 0);
		if arg.is_undefined() { None } else { Some(arg) }
	};
	
	let mut error = None;
	let this = env.new_undefined();
	let zero = env.new_number(0f64);
	
	values.sort_by(|x, y| {
		// Fast escape if we're in error mode.
		if error.is_some() {
			return Ordering::Equal;
		}
		
		if x.is_none() && y.is_none() {
			Ordering::Equal
		} else if x.is_none() {
			Ordering::Greater
		} else if y.is_none() {
			Ordering::Less
		} else {
			let x = x.unwrap();
			let y = y.unwrap();
			
			if x.is_undefined() && y.is_undefined() {
				Ordering::Equal
			} else if x.is_undefined() {
				Ordering::Greater
			} else if y.is_undefined() {
				Ordering::Less
			} else if let Some(compare_fn) = compare_fn {
				if !compare_fn.is_callable(env) {
					error = Some(JsError::new_type(env, ::errors::TYPE_NOT_A_FUNCTION));
					Ordering::Equal
				} else {
					let result = local_try!(compare_fn.call(env, this, vec![x, y], false), error);
					
					if local_try!(env.compare_lt(result, zero), error) {
						Ordering::Less
					} else if local_try!(env.compare_gt(result, zero), error) {
						Ordering::Greater
					} else {
						Ordering::Equal
					}
				}
			} else {
				let x_string = local_try!(x.to_string(env), error);
				let y_string = local_try!(y.to_string(env), error);
				let x_string = x_string.as_value(env);
				let y_string = y_string.as_value(env);
				
				if local_try!(env.compare_lt(x_string, y_string), error) {
					Ordering::Less
				} else if local_try!(env.compare_gt(x_string, y_string), error) {
					Ordering::Greater
				} else {
					Ordering::Equal
				}
			}
		}
	});
	
	if let Some(error) = error {
		return Err(error);
	}
	
	for i in 0..values.len() as usize {
		if let Some(value) = values[i] {
			try!(obj.put(env, Name::from_index(i), value, true));
		} else {
			try!(obj.delete(env, Name::from_index(i), true));
		}
	}
	
	Ok(obj)
}

	let len = try!(len_val.to_uint32(env)) as isize;
		try!(args.arg(env, 0).to_integer(env)) as isize
	let actual_start = if relative_start < 0 {
		cmp::max(relative_start + len, 0) as usize
		if actual_start < 0f64 { 0 } else { actual_start as usize }
		cmp::min(relative_start, len) as usize
		try!(args.arg(env, 1).to_integer(env)) as isize
	let delete_count = if delete_count < 0f64 { 0f64 } else { delete_count };
	let actual_delete_count = if delete_count < (len - actual_start) as f64 {
		delete_count as usize
		0
// 15.4.4.13 Array.prototype.unshift ( [ item1 [ , item2 [ , … ] ] ] )
pub fn Array_unshift(env: &mut JsEnv, _mode: JsFnMode, args: JsArgs) -> JsResult<Local<JsValue>> {
	let mut array = try!(args.this(env).to_object(env));
	let len_val = try!(array.get(env, name::LENGTH));
	let len = try!(len_val.to_uint32(env)) as usize;
	
	let arg_count = args.argc;
	let mut k = len;
	while k > 0 {
		let from = Name::from_index(k - 1);
		let to = Name::from_index(k + arg_count - 1);
		let from_present = array.has_property(env, from);
		if from_present {
			let from_value = try!(array.get(env, from));
			try!(array.put(env, to, from_value, true));
		} else {
			try!(array.delete(env, to, true));
		}
		
		k -= 1;
	}
	
	for i in 0..args.argc {
		let arg = args.arg(env, i);
		try!(array.put(env, Name::from_index(i), arg, true));
	}
	
	let length = env.new_number((len + arg_count) as f64);
	try!(array.put(env, name::LENGTH, length, true));
	
	Ok(length)
}

// 15.4.4.14 Array.prototype.indexOf ( searchElement [ , fromIndex ] )
pub fn Array_indexOf(env: &mut JsEnv, _mode: JsFnMode, args: JsArgs) -> JsResult<Local<JsValue>> {
	let array = try!(args.this(env).to_object(env));
	let len_value = try!(array.get(env, name::LENGTH));
	let len = try!(len_value.to_uint32(env)) as usize;
	
	let search_element = args.arg(env, 0);
	
	let result = if len == 0 {
		-1
	} else {
		let n = if args.argc > 1 {
			try!(args.arg(env, 1).to_integer(env))
		} else {
			0f64
		};
		
		if n >= len as f64 {
			-1
		} else {
			let k = if n >= 0f64 {
				n as usize
			} else {
				let k = len as f64 - n.abs();
				if k < 0f64 { 0 } else { k as usize }
			};
			
			for i in k..len {
				let present = array.has_property(env, Name::from_index(i));
				if present {
					let element = try!(array.get(env, Name::from_index(i)));
					if env.strict_eq(search_element, element) {
						return Ok(env.new_number(i as f64));
					}
				}
			}
			
			-1
		}
	};
	
	Ok(env.new_number(result as f64))
}

// 15.4.4.15 Array.prototype.lastIndexOf ( searchElement [ , fromIndex ] )
pub fn Array_lastIndexOf(env: &mut JsEnv, _mode: JsFnMode, args: JsArgs) -> JsResult<Local<JsValue>> {
	let array = try!(args.this(env).to_object(env));
	let len_value = try!(array.get(env, name::LENGTH));
	let len = try!(len_value.to_uint32(env)) as usize;
	
	let result = if len == 0 {
		-1
	} else {
		let n = if args.argc > 1 {
			try!(args.arg(env, 1).to_integer(env))
		} else {
			(len - 1) as f64
		};
		
		let mut k = if n >= 0f64 {
			if n > (len - 1) as f64 { len - 1 } else { n as usize }
		} else {
			let n = n.abs();
			if n > len as f64 {
				return Ok(env.new_number(-1f64));
			} else {
				len - n as usize
			}
		};
		
		let search_element = args.arg(env, 0);
		
		loop {
			let k_present = array.has_property(env, Name::from_index(k));
			if k_present {
				let element_k = try!(array.get(env, Name::from_index(k)));
				if env.strict_eq(search_element, element_k) {
					return Ok(env.new_number(k as f64));
				}
			}
			
			if k == 0 {
				break
			}
			
			k -= 1;
		}
		
		-1
	};
	
	Ok(env.new_number(result as f64))
}

	let this_arg = args.arg(env, 0);
	let this_arg = args.arg(env, 0);
		if k_present {
			let mapped_value = try!(callback_fn.call(
			));
		} else {
			env.new_undefined()
		};
			
	let this_arg = args.arg(env, 0);
	let accumulator = if args.argc > 1 {
			let accumulator = try!(callback_fn.call(
				vec![k_value, k, array],
	let accumulator = if args.argc > 1 {
		while accumulator.is_none() && k >= len {
	while k >= len {
			let accumulator = try!(callback_fn.call(
				vec![k_value, k, array],
// 15.4.3.2 Array.isArray ( arg )
pub fn Array_isArray(env: &mut JsEnv, _mode: JsFnMode, args: JsArgs) -> JsResult<Local<JsValue>> {
	let arg = args.arg(env, 0);
	
	let result = if arg.ty() != JsType::Object {
		false
	} else {
		if arg.class(env) == Some(name::ARRAY_CLASS) {
			true
		} else if arg.get_ptr() == env.array_prototype.as_ptr() {
			// TODO: This is not conform the specs (the specs state that only the [[Class]]
			// should be checked) but the ECMA test suite tests for this and Chrome
			// passes this test.
			true
		} else {
			false
		}
	};
	
	Ok(env.new_bool(result))
}
